File: BaekJoon/Solutions/18258_큐_2/solution.js
/*
  - 문제 Link : https://www.acmicpc.net/problem/18258
*/

const fs = require("fs");
const { exit } = require("process");
const filePath = process.env.RUNNING_ON === "local" ? "./stdin" : "/dev/stdin";
const input = fs.readFileSync(filePath).toString().split("\n");

class QNode {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.size = 0;
    this.head = null;
    this.tail = null;
  }

  push(value) {
    let newNode = new QNode(value);

    if (!this.head) {
      this.head = newNode;
      this.head.next = this.tail;
    } else {
      this.tail.next = newNode;
    }

    this.tail = newNode;
    this.size += 1;
  }

  pop() {
    if (this.size === 0) {
      return -1;
    }

    let value = this.head.value;
    this.head = this.head.next;

    this.size -= 1;

    return value;
  }

  front() {
    if (this.size < 1) {
      return -1;
    }

    return this.head.value;
  }

  back() {
    if (this.size < 1) {
      return -1;
    }

    return this.tail.value;
  }

  empty() {
    return this.size === 0 ? 1 : 0;
  }
}

let queue = new Queue();

function solution(data) {
  const [cmd, val] = data.split(" ");

  switch (cmd) {
    case "push":
      queue.push(parseInt(val));
      return;
    case "pop":
      return queue.pop();
    case "size":
      return queue.size;
    case "empty":
      return queue.empty();
    case "front":
      return queue.front();
    case "back":
      return queue.back();
  }
}

function main(input) {
  const [n, ...data] = input;

  let answer = [];

  for (let i = 0; i < n; i++) {
    let result = solution(data[i]);
    if (result !== undefined) {
      answer.push(result + "");
    }
  }
  console.log(answer.join("\n"));
}

main(input);

File: BaekJoon/Solutions/18258_큐_2/solution.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

class Queue
{
private:
    int *_data;
    int _header;
    int _tail;
    int _size;
    int _maxSize;

public:
    Queue(int size)
    {
        _maxSize = size;
        _data = new int[_maxSize];
        _size = 0;
        _header = 0;
        _tail = 0;
    };

    void push(int x)
    {
        _size += 1;
        _data[_tail] = x;
        _tail = (_tail + 1) % _maxSize;
    }

    int pop()
    {
        if (empty())
            return -1;

        int result = _data[_header];
        _size -= 1;

        _header = (_header + 1) % _maxSize;

        return result;
    }

    int size()
    {
        return _size;
    }

    int empty()
    {
        return _size == 0;
    }

    int front()
    {
        return empty() ? -1 : _data[_header];
    }

    int back()
    {
        return empty() ? -1 : _data[_tail - 1];
    }

    ~Queue()
    {
        delete _data;
    };
};

int main()
{
    cin.tie(nullptr);
    ios::sync_with_stdio(false);

    int N;
    cin >> N;

    Queue q = Queue(N);

    for (int i = 0; i < N; i++)
    {
        string cmd;
        cin >> cmd;

        if (cmd == "push")
        {
            int X;
            cin >> X;
            q.push(X);
        }
        else if (cmd == "pop")
        {
            cout << q.pop() << "\n";
        }
        else if (cmd == "size")
        {
            cout << q.size() << "\n";
        }
        else if (cmd == "empty")
        {
            cout << (q.empty() ? 1 : 0) << "\n";
        }
        else if (cmd == "front")
        {
            cout << q.front() << "\n";
        }
        else if (cmd == "back")
        {
            cout << q.back() << "\n";
        }
    }

    return 0;
}